// @strapi/plugin-color-picker/dist/admin/index.mjs
import { ColorPickerIcon } from './components/ColorPickerIcon.mjs';
import { pluginId } from './pluginId.mjs';
import { getTrad } from './utils/getTrad.mjs';
import { prefixPluginTranslations } from './utils/prefixPluginTranslations.mjs';

function __variableDynamicImportRuntime1__(path) {
  switch (path) {
    case './translations/cs.json': return import('./translations/cs.json.mjs');
    case './translations/en.json': return import('./translations/en.json.mjs');
    case './translations/ru.json': return import('./translations/ru.json.mjs');
    case './translations/sv.json': return import('./translations/sv.json.mjs');
    case './translations/tr.json': return import('./translations/tr.json.mjs');
    case './translations/uk.json': return import('./translations/uk.json.mjs');
    case './translations/zh.json': return import('./translations/zh.json.mjs');
    default: return new Promise(function(resolve, reject) {
      (typeof queueMicrotask === 'function' ? queueMicrotask : setTimeout)(
        reject.bind(null, new Error("Unknown variable dynamic import: " + path))
      );
    })
   }
 }
// eslint-disable-next-line import/no-default-export
var index = {
    /**
   * TODO: we need to have the type for StrapiApp done from `@strapi/admin` package.
   */ // eslint-disable-next-line @typescript-eslint/no-explicit-any
    register (app) {
        app.customFields.register({
            name: 'color',
            pluginId: 'color-picker',
            type: 'string',
            icon: ColorPickerIcon,
            intlLabel: {
                id: getTrad('color-picker.label'),
                defaultMessage: 'Color'
            },
            intlDescription: {
                id: getTrad('color-picker.description'),
                defaultMessage: 'Select any color'
            },
            components: {
                Input: async ()=>import('./components/ColorPickerInput.mjs').then((module)=>({
                            default: module.ColorPickerInput
                        }))
            },
            options: {
                advanced: [
                    {
                        intlLabel: {
                            id: getTrad('color-picker.options.advanced.regex'),
                            defaultMessage: 'RegExp pattern'
                        },
                        name: 'regex',
                        type: 'text',
                        defaultValue: '^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$',
                        description: {
                            id: getTrad('color-picker.options.advanced.regex.description'),
                            defaultMessage: 'The text of the regular expression'
                        }
                    },
                    {
                        sectionTitle: {
                            id: 'global.settings',
                            defaultMessage: 'Settings'
                        },
                        items: [
                            {
                                name: 'required',
                                type: 'checkbox',
                                intlLabel: {
                                    id: getTrad('color-picker.options.advanced.requiredField'),
                                    defaultMessage: 'Required field'
                                },
                                description: {
                                    id: getTrad('color-picker.options.advanced.requiredField.description'),
                                    defaultMessage: "You won't be able to create an entry if this field is empty"
                                }
                            }
                        ]
                    }
                ]
            }
        });
    },
    async registerTrads ({ locales }) {
        const importedTrads = await Promise.all(locales.map((locale)=>{
            return __variableDynamicImportRuntime1__(`./translations/${locale}.json`).then(({ default: data })=>{
                return {
                    data: prefixPluginTranslations(data, pluginId),
                    locale
                };
            }).catch(()=>{
                return {
                    data: {},
                    locale
                };
            });
        }));
        return Promise.resolve(importedTrads);
    }
};

export { index as default };
