// @strapi/admin/dist/admin/admin/src/components/UnstableGuidedTour/Tours.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var styled = require('styled-components');
var admin = require('../../services/admin.js');
var Context = require('./Context.js');
var Step = require('./Step.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

/* -------------------------------------------------------------------------------------------------
 * Tours
 * -----------------------------------------------------------------------------------------------*/ const tours = {
    TEST: createTour('TEST', [
        {
            name: 'Introduction',
            content: (Step)=>/*#__PURE__*/ jsxRuntime.jsxs(Step.Root, {
                    sideOffset: -36,
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(Step.Title, {
                            id: "tours.contentManager.Introduction.title",
                            defaultMessage: "Content manager"
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(Step.Content, {
                            id: "tours.contentManager.Introduction.content",
                            defaultMessage: "Create and manage content from your collection types and single types."
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(Step.Actions, {
                            showSkip: true
                        })
                    ]
                })
        },
        {
            name: 'Done',
            requiredActions: [
                'didCreateApiToken'
            ],
            content: (Step)=>/*#__PURE__*/ jsxRuntime.jsxs(Step.Root, {
                    align: "start",
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(Step.Title, {
                            id: "tours.contentManager.CreateEntry.title",
                            defaultMessage: "Create entry"
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(Step.Content, {
                            id: "tours.contentManager.CreateEntry.content",
                            defaultMessage: "Click this button to create an entry"
                        }),
                        /*#__PURE__*/ jsxRuntime.jsx(Step.Actions, {})
                    ]
                })
        }
    ])
};
const GuidedTourOverlay = styled.styled(designSystem.Box)`
  position: fixed;
  inset: 0;
  background-color: rgba(50, 50, 77, 0.2);
  z-index: 10;
`;
const UnstableGuidedTourTooltip = ({ children, content, tourName, step, requiredActions })=>{
    const { data: guidedTourMeta } = admin.useGetGuidedTourMetaQuery();
    const state = Context.unstableUseGuidedTour('UnstableGuidedTourTooltip', (s)=>s.state);
    const dispatch = Context.unstableUseGuidedTour('UnstableGuidedTourTooltip', (s)=>s.dispatch);
    const Step$1 = React__namespace.useMemo(()=>Step.createStepComponents(tourName), [
        tourName
    ]);
    const isCurrentStep = state.tours[tourName].currentStep === step;
    const hasCompletedRequiredActions = requiredActions?.every((action)=>{
        return guidedTourMeta?.data?.completedActions.includes(action);
    }) ?? true;
    const hasFutureFlag = window.strapi.future.isEnabled('unstableGuidedTour');
    const isEnabled = guidedTourMeta?.data?.isFirstSuperAdminUser && !state.tours[tourName].isCompleted && hasFutureFlag;
    const isPopoverOpen = isEnabled && isCurrentStep && hasCompletedRequiredActions;
    // Lock the scroll
    React__namespace.useEffect(()=>{
        if (!isPopoverOpen) return;
        const originalStyle = window.getComputedStyle(document.body).overflow;
        document.body.style.overflow = 'hidden';
        return ()=>{
            document.body.style.overflow = originalStyle;
        };
    }, [
        isPopoverOpen
    ]);
    return /*#__PURE__*/ jsxRuntime.jsxs(jsxRuntime.Fragment, {
        children: [
            isPopoverOpen && /*#__PURE__*/ jsxRuntime.jsx(GuidedTourOverlay, {}),
            /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Popover.Root, {
                open: isPopoverOpen,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Popover.Anchor, {
                        children: children
                    }),
                    content(Step$1, {
                        state,
                        dispatch
                    })
                ]
            })
        ]
    });
};
function createTour(tourName, steps) {
    const tour = steps.reduce((acc, step, index)=>{
        if (step.name in acc) {
            throw Error(`The tour: ${tourName} with step: ${step.name} has already been registered`);
        }
        acc[step.name] = ({ children })=>/*#__PURE__*/ jsxRuntime.jsx(UnstableGuidedTourTooltip, {
                tourName: tourName,
                step: index,
                content: step.content,
                requiredActions: step.requiredActions,
                children: children
            });
        return acc;
    }, {});
    return tour;
}

exports.GuidedTourOverlay = GuidedTourOverlay;
exports.tours = tours;
