// @strapi/plugin-color-picker/dist/admin/components/ColorPickerInput.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var admin = require('@strapi/strapi/admin');
var reactColorful = require('react-colorful');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var getTrad = require('../utils/getTrad.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);

const ColorPreview = styledComponents.styled.div`
  border-radius: 50%;
  width: 20px;
  height: 20px;
  margin-right: 10px;
  background-color: ${(props)=>props.color};
  border: 1px solid rgba(0, 0, 0, 0.1);
`;
const ColorPicker = styledComponents.styled(reactColorful.HexColorPicker)`
  && {
    width: 100%;
    aspect-ratio: 1.5;
  }

  .react-colorful__pointer {
    width: ${({ theme })=>theme.spaces[3]};
    height: ${({ theme })=>theme.spaces[3]};
  }

  .react-colorful__saturation {
    border-radius: ${({ theme })=>theme.spaces[1]};
    border-bottom: none;
  }

  .react-colorful__hue {
    border-radius: 10px;
    height: ${({ theme })=>theme.spaces[3]};
    margin-top: ${({ theme })=>theme.spaces[2]};
  }
`;
const ColorPickerToggle = styledComponents.styled(designSystem.Button)`
  & > span {
    display: flex;
    justify-content: space-between;
    align-items: center;
    width: 100%;
  }

  svg {
    width: ${({ theme })=>theme.spaces[2]};
    height: ${({ theme })=>theme.spaces[2]};
  }

  svg > path {
    fill: ${({ theme })=>theme.colors.neutral500};
    justify-self: flex-end;
  }
`;
const ColorPickerPopover = styledComponents.styled(designSystem.Popover.Content)`
  padding: ${({ theme })=>theme.spaces[2]};
  min-height: 270px;
`;
const ColorPickerInput = /*#__PURE__*/ React__namespace.forwardRef(({ hint, disabled, labelAction, label, name, required, ...props }, forwardedRef)=>{
    const [showColorPicker, setShowColorPicker] = React__namespace.useState(false);
    const colorPickerButtonRef = React__namespace.useRef(null);
    const { formatMessage } = reactIntl.useIntl();
    const field = admin.useField(name);
    const color = field.value ?? '#000000';
    const composedRefs = designSystem.useComposedRefs(forwardedRef, colorPickerButtonRef);
    return /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Root, {
        name: name,
        id: name,
        error: field.error,
        hint: hint,
        required: required,
        children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
            direction: "column",
            alignItems: "stretch",
            gap: 1,
            children: [
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                    action: labelAction,
                    children: label
                }),
                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Popover.Root, {
                    onOpenChange: setShowColorPicker,
                    children: [
                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Popover.Trigger, {
                            children: /*#__PURE__*/ jsxRuntime.jsxs(ColorPickerToggle, {
                                ref: composedRefs,
                                "aria-label": formatMessage({
                                    id: getTrad.getTrad('color-picker.toggle.aria-label'),
                                    defaultMessage: 'Color picker toggle'
                                }),
                                "aria-controls": "color-picker-value",
                                "aria-haspopup": "dialog",
                                "aria-expanded": showColorPicker,
                                "aria-disabled": disabled,
                                disabled: disabled,
                                variant: "tertiary",
                                size: "L",
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                        children: [
                                            /*#__PURE__*/ jsxRuntime.jsx(ColorPreview, {
                                                color: color
                                            }),
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                style: {
                                                    textTransform: 'uppercase'
                                                },
                                                textColor: field.value ? undefined : 'neutral600',
                                                variant: "omega",
                                                children: color
                                            })
                                        ]
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsx(icons.CaretDown, {
                                        "aria-hidden": true
                                    })
                                ]
                            })
                        }),
                        /*#__PURE__*/ jsxRuntime.jsxs(ColorPickerPopover, {
                            sideOffset: 4,
                            children: [
                                /*#__PURE__*/ jsxRuntime.jsx(ColorPicker, {
                                    color: color,
                                    onChange: (hexValue)=>field.onChange(name, hexValue)
                                }),
                                /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                    paddingTop: 3,
                                    paddingLeft: 4,
                                    justifyContent: "flex-end",
                                    children: [
                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                                            paddingRight: 2,
                                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                                variant: "omega",
                                                tag: "label",
                                                textColor: "neutral600",
                                                children: formatMessage({
                                                    id: getTrad.getTrad('color-picker.input.format'),
                                                    defaultMessage: 'HEX'
                                                })
                                            })
                                        }),
                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Root, {
                                            children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Input, {
                                                "aria-label": formatMessage({
                                                    id: getTrad.getTrad('color-picker.input.aria-label'),
                                                    defaultMessage: 'Color picker input'
                                                }),
                                                style: {
                                                    textTransform: 'uppercase'
                                                },
                                                name: name,
                                                defaultValue: color,
                                                placeholder: "#000000",
                                                onChange: field.onChange,
                                                ...props
                                            })
                                        })
                                    ]
                                })
                            ]
                        })
                    ]
                }),
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Hint, {}),
                /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Error, {})
            ]
        })
    });
});

exports.ColorPickerInput = ColorPickerInput;
