// @strapi/plugin-documentation/dist/admin/components/SettingsForm.js
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var designSystem = require('@strapi/design-system');
var icons = require('@strapi/icons');
var admin = require('@strapi/strapi/admin');
var formik = require('formik');
var reactIntl = require('react-intl');
var styledComponents = require('styled-components');
var yup = require('yup');
var constants = require('../constants.js');
var getTrad = require('../utils/getTrad.js');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespaceDefault(React);
var yup__namespace = /*#__PURE__*/_interopNamespaceDefault(yup);

const schema = yup__namespace.object().shape({
    restrictedAccess: yup__namespace.boolean(),
    password: yup__namespace.string().when('restrictedAccess', (value, initSchema)=>{
        return value ? initSchema.required(admin.translatedErrors.required.id).min(8).matches(/[a-z]/, 'components.Input.error.contain.lowercase').matches(/[A-Z]/, 'components.Input.error.contain.uppercase').matches(/\d/, 'components.Input.error.contain.number') : initSchema;
    })
});
const FieldActionWrapper = styledComponents.styled(designSystem.Field.Action)`
  svg {
    height: 1.6rem;
    width: 1.6rem;
    path {
      fill: ${({ theme })=>theme.colors.neutral600};
    }
  }
`;
const SettingsForm = ({ data, onSubmit })=>{
    const { formatMessage } = reactIntl.useIntl();
    const [passwordShown, setPasswordShown] = React__namespace.useState(false);
    const { allowedActions } = admin.useRBAC(constants.PERMISSIONS);
    return /*#__PURE__*/ jsxRuntime.jsx(formik.Formik, {
        enableReinitialize: true,
        initialValues: {
            restrictedAccess: data?.documentationAccess.restrictedAccess || false,
            password: ''
        },
        onSubmit: onSubmit,
        validationSchema: schema,
        children: ({ handleSubmit, values, handleChange, errors, setFieldTouched, setFieldValue, setFieldError, dirty })=>{
            return /*#__PURE__*/ jsxRuntime.jsxs(formik.Form, {
                noValidate: true,
                onSubmit: handleSubmit,
                children: [
                    /*#__PURE__*/ jsxRuntime.jsx(admin.Layouts.Header, {
                        title: formatMessage({
                            id: getTrad.getTrad('plugin.name'),
                            defaultMessage: 'Documentation'
                        }),
                        subtitle: formatMessage({
                            id: getTrad.getTrad('pages.SettingsPage.header.description'),
                            defaultMessage: 'Configure the documentation plugin'
                        }),
                        primaryAction: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Button, {
                            type: "submit",
                            startIcon: /*#__PURE__*/ jsxRuntime.jsx(icons.Check, {}),
                            disabled: !dirty && allowedActions.canUpdate,
                            children: formatMessage({
                                id: getTrad.getTrad('pages.SettingsPage.Button.save'),
                                defaultMessage: 'Save'
                            })
                        })
                    }),
                    /*#__PURE__*/ jsxRuntime.jsx(admin.Layouts.Content, {
                        children: /*#__PURE__*/ jsxRuntime.jsx(designSystem.Box, {
                            background: "neutral0",
                            hasRadius: true,
                            shadow: "filterShadow",
                            paddingTop: 6,
                            paddingBottom: 6,
                            paddingLeft: 7,
                            paddingRight: 7,
                            children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Flex, {
                                direction: "column",
                                alignItems: "stretch",
                                gap: 4,
                                children: [
                                    /*#__PURE__*/ jsxRuntime.jsx(designSystem.Typography, {
                                        variant: "delta",
                                        tag: "h2",
                                        children: formatMessage({
                                            id: 'global.settings',
                                            defaultMessage: 'Settings'
                                        })
                                    }),
                                    /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Grid.Root, {
                                        gap: 4,
                                        children: [
                                            /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                                                col: 6,
                                                s: 12,
                                                direction: "column",
                                                alignItems: "stretch",
                                                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
                                                    name: "restrictedAccess",
                                                    hint: formatMessage({
                                                        id: getTrad.getTrad('pages.SettingsPage.toggle.hint'),
                                                        defaultMessage: 'Make the documentation endpoint private'
                                                    }),
                                                    children: [
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                                                            children: formatMessage({
                                                                id: getTrad.getTrad('pages.SettingsPage.toggle.label'),
                                                                defaultMessage: 'Restricted Access'
                                                            })
                                                        }),
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Toggle, {
                                                            checked: values.restrictedAccess,
                                                            onChange: ()=>{
                                                                if (values.restrictedAccess === true) {
                                                                    setFieldValue('password', '', false);
                                                                    setFieldTouched('password', false, false);
                                                                    setFieldError('password', undefined);
                                                                }
                                                                setFieldValue('restrictedAccess', !values.restrictedAccess, false);
                                                            },
                                                            onLabel: "On",
                                                            offLabel: "Off"
                                                        }),
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Hint, {})
                                                    ]
                                                })
                                            }),
                                            values.restrictedAccess && /*#__PURE__*/ jsxRuntime.jsx(designSystem.Grid.Item, {
                                                col: 6,
                                                s: 12,
                                                direction: "column",
                                                alignItems: "stretch",
                                                children: /*#__PURE__*/ jsxRuntime.jsxs(designSystem.Field.Root, {
                                                    name: "password",
                                                    error: errors.password ? formatMessage({
                                                        id: errors.password,
                                                        defaultMessage: errors.password
                                                    }) : undefined,
                                                    children: [
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Label, {
                                                            children: formatMessage({
                                                                id: 'global.password',
                                                                defaultMessage: 'Password'
                                                            })
                                                        }),
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.TextInput, {
                                                            placeholder: "**********",
                                                            type: passwordShown ? 'text' : 'password',
                                                            value: values.password,
                                                            onChange: handleChange,
                                                            endAction: /*#__PURE__*/ jsxRuntime.jsx(FieldActionWrapper, {
                                                                onClick: (e)=>{
                                                                    e.stopPropagation();
                                                                    setPasswordShown((prev)=>!prev);
                                                                },
                                                                label: formatMessage(passwordShown ? {
                                                                    id: 'Auth.form.password.show-password',
                                                                    defaultMessage: 'Show password'
                                                                } : {
                                                                    id: 'Auth.form.password.hide-password',
                                                                    defaultMessage: 'Hide password'
                                                                }),
                                                                children: passwordShown ? /*#__PURE__*/ jsxRuntime.jsx(icons.Eye, {}) : /*#__PURE__*/ jsxRuntime.jsx(icons.EyeStriked, {})
                                                            })
                                                        }),
                                                        /*#__PURE__*/ jsxRuntime.jsx(designSystem.Field.Error, {})
                                                    ]
                                                })
                                            })
                                        ]
                                    })
                                ]
                            })
                        })
                    })
                ]
            });
        }
    });
};

exports.SettingsForm = SettingsForm;
